### ISSUE TITLE
Support for "dynamic" entity models

### ISSUE THREAD

**Issue by divega** [dotnet/efcore#2282]
The goal of this issue is to track and rationalize various levels or flavors of "dynamic behavior" we could attain with EF in the future. 

The initial list is based on digesting information provided in various bugs such as #1513 and #2141 as well as conversations with various teams and previous research on feature gaps with other persistence frameworks. 

Note that none of this is planned for EF Core 1.0.0 and before we start bringing dynamic features into EF we should probably open more issues and prioritize each individual feature independently. But at least for now should any new "dynamic" scenario be brought to our attention by the community or should our understanding of them improve we can just keep this list up to date:
- [X] ~**Shadow objects:** (already tracked individually as #749) This is about having certain entities or complex types defined in the model that do not get materialized as user objects but are just tracked as "state" in the DbContext instance. This is useful for such scenarios as link tables in many-to-many associations or any other situation in which there is an entire entity that contains information that EF needs to track but that the user has chosen not to include in the object model.~
- [X] **Property bags:** Currently the EF Core stack assumes a distinct CLR type is mapped to each distinct entity type (and eventually for complex types) in the entity model and a CLR property for each entity property (minus any shadow property). This could be changed to enable EF to materialize and track simple property bags (e.g. any object that implements `IDictionary<string, object>` or contains an indexer property or any other mechanism to get and set values based on a property name). Even the name of the entity instance belongs to could become another property in the bag. This feature would make working with entity models that are created programmatically easier because it would remove the need to emit new CLR types for each entity and complex type at runtime. (#9914)
- [ ] **Dynamic aggregates:** Allow the entity type to be determined by the set of contained components (#26123)
- [ ] **Proper dynamic objects:** This refers specifically to supporting entities and complex types in which each specific instance can contain a different set of properties. I anticipate that the most likely path for us to support proper dynamic objects mapped to relational databases could be to serialize the dynamic part of the object into a BLOB or JSON column instead of shredding it into the individual scalar properties. Querying based on standard properties can be enabled and optimized through computed columns and indexing as described in the comments of #2141. Even within this category there are several alternative levels of support which will probably pose different challenges:
   1. Only scalar properties can differ vs. relationships can also differ
   2. Whole dynamic objects vs. "open types" (i.e. dynamic properties are added at runtime to a set of predefined properties). 
   3. Property bag representation vs. actual dynamic/ExpandoObject CLR types (the latter can lead to nicer syntax but might be impossible to use in LINQ).

**Comment by divega** [comment]
Moving this straight to the backlog.

**Comment by AcousticGuitar** [comment]
Hello, just a thought on this, instead of DB first there could be a JSON or even XML first equivalent.

**Comment by natemcmaster** [comment]
This would help with materializing `Annotatable` objects. Currently we can get close by using shadow properties and then doing a post-materialization fixup. 

```c#
var post = context.Posts.First();
var entry = context.Entry(post);
var isDraft = entry.Property("IsDraft").CurrentValue as bool;
post.AddAnnotation("IsDraft", isDraft);
```

Of course this has caveats: AFAIK the only way to get a shadow property value after materialization is to use change tracking. This means to `AsNoTracking()` cannot be used if the shadow property is required outside of query.

**Comment by dmitry-a-morozov** [comment]
Are there plans to make a progress on this feature request? 
This will enable implementation of F# erased type providers for EF Core.

**Comment by ajcvickers** [comment]
Triage: Consider scenario described here when tackling this issue: #10555

**Comment by weitzhandler** [comment]
Dynamic functionality is very essential in almost any growing app, especially LoB apps or apps that grow quickly and dynamically.

Let me just demonstrate with a couple of verbal examples using a `Contact` class.
 - A list of values or tuples, for example a list of phones, where each phone has a label (Home, Fax, or whatever), so we should be able to search using EF:
```c#
var number = CleanNumber(number); //remove formatting
dbContext.Contacts
  .Where(c => 
    c.Phones.Where(p => p.Number.Contains(number));
```
Remember that we don't want to predefine the phone shape in the DB (only in the EF back-end).
 - A list of dynamic types. For instance I don't want to define the fields in my `Contact` entity at all, instead I need them to be all dynamic, for example, I'd sometimes want `Contact` to have a person name, whereas sometimes a company name, in many other scenarios the user would want to add custom fields.
If it's supposed to be a dynamic app so that the user is the one that defines which columns should be there, after the data is stored in the DB, we still obviously still want to let the user search all contacts that have a company name _x_ without having any of this defined in the server.
We might want to store schema data externally, but that's already not up to the DB but to the business layer.
From EF's point of view, we should be able when configuring the model, to define some fields, collections, or entities, as dynamic data, which feel as natural columns from the front-ends point of view, but will be transformed to a dynamic means when stored to the DB.

Let's take Microsoft SQL Server as our DB provider. It provides out the box support for that using the XML 
datatype. We could transform all the data fields that we defined as dynamic to be stored and manipulated as XML.

@divega have you been thinking about XML as a means for dynamic data support in MSSQL?

**Comment by smitpatel** [comment]
@weitzhandler - I was reading this thread. While dynamic entity types has a wide scope, I believe the scenario you posted could possibly work in current system already.
Tuple/list may be slightly difficult, but if you are using Dictionary/Array and using JSON on server side to store, you could value converter to store your data. And for querying part use DbFunction to access data on server and get queries translated. It also covers when you don't have LabelNames defined and you want to use dynamic "key".

**Comment by weitzhandler** [comment]
Please check out the [Impatient](https://github.com/tuespetre/Impatient) repo, which relates here.

**Comment by ajcvickers** [comment]
Note from triage: remaining work here tracked by related issues.
